File: server/src/services/documentTypeDetector.js
// Detects "the user is asking for a standalone document" independent of
// file attachment — taskRouter.js's document_analysis/resume_analysis only
// fire when a file is already attached, which is the wrong signal for
// "write me an implementation plan" (no file involved at all). Regex, not
// an LLM call, so this adds zero latency to either chat path.
const DOC_VERB_RE = /\b(write|create|draft|generate|prepare|build|design|make)\b/i;

const DOC_TYPE_PATTERNS = [
  [/implementation plan/i, 'implementation_plan', 'Implementation Plan'],
  [/project report/i, 'project_report', 'Project Report'],
  [/research paper/i, 'research_paper', 'Research Paper'],
  [/\bsrs\b|software requirements? specification/i, 'srs', 'Software Requirements Specification'],
  [/\bresume\b|\bcv\b/i, 'resume', 'Resume'],
  [/cover letter/i, 'cover_letter', 'Cover Letter'],
  [/business proposal/i, 'business_proposal', 'Business Proposal'],
  [/meeting minutes/i, 'meeting_minutes', 'Meeting Minutes'],
  [/api documentation/i, 'api_documentation', 'API Documentation'],
  [/technical documentation/i, 'technical_documentation', 'Technical Documentation'],
  [/\bassignment\b/i, 'assignment', 'Assignment'],
  [/white ?paper/i, 'white_paper', 'White Paper'],
  [/\bdocumentation\b/i, 'documentation', 'Documentation']
];

export function detectDocumentRequest(query) {
  if (!query || !DOC_VERB_RE.test(query)) return null;
  for (const [re, type, label] of DOC_TYPE_PATTERNS) {
    if (re.test(query)) return { type, label };
  }
  return null;
}

// Strips markdown syntax for a short preview line — same intent as
// DocumentCard's existing inline preview, used here for the `summary` field
// stored alongside the full content.
export function buildSummary(answer, maxLen = 160) {
  const stripped = (answer || '').replace(/[#*_`>-]/g, ' ').replace(/\s+/g, ' ').trim();
  return stripped.length > maxLen ? `${stripped.slice(0, maxLen)}…` : stripped;
}

File: server/app/services/document_type_detector.py
import re

DOC_VERB_RE = re.compile(r"\b(write|create|draft|generate|prepare|build|design|make)\b", re.I)

DOC_TYPE_PATTERNS = [
    (re.compile(r"implementation plan", re.I), "implementation_plan", "Implementation Plan"),
    (re.compile(r"project report", re.I), "project_report", "Project Report"),
    (re.compile(r"research paper", re.I), "research_paper", "Research Paper"),
    (re.compile(r"\bsrs\b|software requirements? specification", re.I), "srs", "Software Requirements Specification"),
    (re.compile(r"\bresume\b|\bcv\b", re.I), "resume", "Resume"),
    (re.compile(r"cover letter", re.I), "cover_letter", "Cover Letter"),
    (re.compile(r"business proposal", re.I), "business_proposal", "Business Proposal"),
    (re.compile(r"meeting minutes", re.I), "meeting_minutes", "Meeting Minutes"),
    (re.compile(r"api documentation", re.I), "api_documentation", "API Documentation"),
    (re.compile(r"technical documentation", re.I), "technical_documentation", "Technical Documentation"),
    (re.compile(r"\bassignment\b", re.I), "assignment", "Assignment"),
    (re.compile(r"white ?paper", re.I), "white_paper", "White Paper"),
    (re.compile(r"\bdocumentation\b", re.I), "documentation", "Documentation"),
]


def detect_document_request(query: str):
    if not query or not DOC_VERB_RE.search(query):
        return None
    for regex, doc_type, label in DOC_TYPE_PATTERNS:
        if regex.search(query):
            return {"type": doc_type, "label": label}
    return None


def build_summary(answer: str, max_len: int = 160) -> str:
    stripped = re.sub(r"\s+", " ", re.sub(r"[#*_`>-]", " ", answer or "")).strip()
    return f"{stripped[:max_len]}…" if len(stripped) > max_len else stripped
